python/ray/tune/syncer.py

```diff
@@ -142,395 +79,408 @@ class SyncConfig:
     """
 
     upload_dir: Optional[str] = None
-    syncer: Optional[str] = "auto"
+    syncer: Optional[Union[str, "Syncer"]] = "auto"
 
     sync_on_checkpoint: bool = True
-    sync_period: int = 300
+    sync_period: int = DEFAULT_SYNC_PERIOD
+
+
+class _BackgroundProcess:
+    def __init__(self, fn: Callable):
+        self._fn = fn
+        self._process = None
+        self._result = {}
+
+    @property
+    def is_running(self):
+        return self._process and self._process.is_alive()
+
+    def start(self, *args, **kwargs):
+        if self.is_running:
+            return False
+
+        self._result = {}
+
+        def entrypoint():
+            try:
+                result = self._fn(*args, **kwargs)
+            except Exception as e:
+                self._result["exception"] = e
+                return
+
+            self._result["result"] = result
+
+        self._process = threading.Thread(target=entrypoint)
+        self._process.start()
+
+    def wait(self):
+        if not self._process:
+            return
+
+        self._process.join()
+        self._process = None
+
+        exception = self._result.get("exception")
+        if exception:
+            raise exception
+
+        result = self._result.get("result")
+
+        self._result = {}
+        return result
 
 
 @DeveloperAPI
-class Syncer:
-    def __init__(self, local_dir: str, remote_dir: str, sync_client: SyncClient = NOOP):
-        """Syncs between two directories with the sync_function.
-
-        Arguments:
-            local_dir: Directory to sync. Uniquely identifies the syncer.
-            remote_dir: Remote directory to sync with.
-            sync_client: Client for syncing between local_dir and
-                remote_dir. Defaults to a Noop.
-        """
-        self._local_dir = os.path.join(local_dir, "") if local_dir else local_dir
-        self._remote_dir = remote_dir
+class Syncer(abc.ABC):
```

Comment:
Overall, I feel like syncing to cloud provides less guarantee than syncing to driver. Is this what we/user want?

Reply:
Why do you feel that way? For syncing to cloud we retry up to three times per default and verbosely log sync failures, so I think we actually do a good job here.

We don't have strong consistency targets right now, but that's true for both syncing mechanisms and something we should work towards to in the future.